cpp/ql/src/experimental/Security/CWE/CWE-193/InvalidPointerDeref.ql

```diff
@@ -0,0 +1,347 @@
+  pai.getRight() = right
+}
+
+/**
+ * Holds if `pai` is non-strictly upper bounded by `sink2 + delta` and `sink1` is the
+ * left operand of the pointer-arithmetic operation.
+ *
+ * For example in,
+ * ```cpp
+ * char* end = p + (size + 1);
+ * ```
+ * We will have:
+ * - `pai` is `p + (size + 1)`,
+ * - `sink1` is `p`
+ * - `sink2` is `size`
+ * - `delta` is `1`.
+ */
+pragma[nomagic]
+predicate pointerAddInstructionHasBounds(
+  PointerAddInstruction pai, DataFlow::Node sink1, Instruction sink2, int delta
+) {
+  exists(Instruction right |
+    pointerAddInstructionHasOperands(pai, sink1.asInstruction(), right) and
+    bounded(right, sink2, delta)
+  )
+}
+
+/**
+ * Holds if `pai` is non-strictly upper bounded by `sink2 + delta` and `sink1` is the
+ * left operand of the pointer-arithmetic operation.
+ *
+ * See `pointerAddInstructionHasBounds` for an example.
+ */
+predicate isSinkImpl(
+  PointerAddInstruction pai, DataFlow::Node sink1, DataFlow::Node sink2, int delta
+) {
+  pointerAddInstructionHasBounds(pai, sink1, sink2.asInstruction(), delta)
+}
+
+/**
+ * Holds if `sink` is a sink for `InvalidPointerToDerefConf` and `i` is a `StoreInstruction` that
+ * writes to an address that non-strictly upper-bounds `sink`, or `i` is a `LoadInstruction` that
+ * reads from an address that non-strictly upper-bounds `sink`.
+ */
+predicate isInvalidPointerDerefSink(DataFlow::Node sink, Instruction i, string operation) {
+  exists(AddressOperand addr, int delta |
+    bounded(addr.getDef(), sink.asInstruction(), delta) and
+    delta >= 0 and
+    i.getAnOperand() = addr
+  |
+    i instanceof StoreInstruction and
+    operation = "write"
+    or
+    i instanceof LoadInstruction and
+    operation = "read"
+  )
+}
+
+/**
+ * A configuration to track flow from a pointer-arithmetic operation found
+ * by `AllocToInvalidPointerConf` to a dereference of the pointer.
+ */
+class InvalidPointerToDerefConf extends DataFlow3::Configuration {
+  InvalidPointerToDerefConf() { this = "InvalidPointerToDerefConf" }
+
+  override predicate isSource(DataFlow::Node source) { invalidPointerToDerefSource(_, source, _) }
+
+  override predicate isSink(DataFlow::Node sink) { isInvalidPointerDerefSink(sink, _, _) }
+}
+
+/**
+ * Holds if `pai` is a pointer-arithmetic operation and `source` is a dataflow node with a
+ * pointer-value that is non-strictly upper bounded by `pai + delta`.
+ *
+ * For example, if `pai` is a pointer-arithmetic operation `p + size` in an expression such
+ * as `(p + size) + 1` and `source` is the node representing `(p + size) + 1`. In this
+ * case `delta` is 1.
+ */
+predicate invalidPointerToDerefSource(
+  PointerArithmeticInstruction pai, DataFlow::Node source, int delta
+) {
+  exists(ProductFlow::Configuration conf, DataFlow::PathNode p, DataFlow::Node sink1 |
+    p.getNode() = sink1 and
+    conf.hasFlowPath(_, _, p, _) and
+    isSinkImpl(pai, sink1, _, _) and
+    bounded(source.asInstruction(), pai, delta) and
+    delta >= 0
```

Comment:
Something feels off here.  I may be wrong, there are a lot of subtleties going on.

1. we never use the state / delta value from the first dataflow config anywhere.  Without this I don't see how we can distinguish cases like:

```
char *buffer1 = malloc(size);
char *ptr1 = buffer1 + size;
*ptr1; // BAD

char *buffer2 = malloc(size + 1);
char *ptr2 = buffer2 + size;
*ptr2; // GOOD
```

2. also I think it might be OK to track a negative delta as long as the sum is positive by the end (dereference), e.g. to find this result:
```
char *p = malloc(size);
char *q = p + size - 1; // seems OK here
char a = *(q + 1); // BAD
```
(I'm not entirely sure we don't get this right, with `(q + 1)` as the midpoint between the two dataflows in this case rather than `p + size - 1`???)

Reply:
> we never use the state / delta value from the first dataflow config anywhere. Without this I don't see how we can distinguish cases like:
> ```cpp
> char *buffer1 = malloc(size);
> char *ptr1 = buffer1 + size;
> *ptr1; // BAD
>
> char *buffer2 = malloc(size + 1);
> char *ptr2 = buffer2 + size;
> *ptr2; // GOOD
> ```
We do actually use the state to precisely distinguish these cases. Take a look at the `isSinkPair` (and `isSourcePair`) definitions in the `AllocToInvalidPointerConf` configuration. This is the source:
```ql
state1 = "" and
hasSize(source1.asConvertedExpr(), source2, state2)
```
and this is the sink:
```ql
state1 = "" and
exists(int delta |
  isSinkImpl(_, sink1, sink2, delta) and
  state2 = delta.toString()
)
```
So in `isSourcePair`, we're using the state to capture the fact that the size is `source2` plus a constant offset stored in `stage2`, and we're using that state in `isSinkPair` to check that the end pointer being computed has the same offset. So in this example:
```cpp
char *buffer2 = malloc(size + 1);
char *ptr2 = buffer2 + size;
*ptr2; // GOOD
```
the state of the second projection of the product dataflow configuration (i.e., `state2`) is `1` because the allocation has size `size + 1`, and then `buffer2 + size` will only be a sink when `state2` is `0`.

Does that make sense?

Reply:
>2. also I think it might be OK to track a negative delta as long as the sum is positive by the end (dereference), e.g. to find this result:
> ```cpp
> char *p = malloc(size);
> char *q = p + size - 1; // seems OK here
> char a = *(q + 1); // BAD
> ```
>
> (I'm not entirely sure we don't get this right, with (q + 1) as the midpoint between the two dataflows in this case rather than p + size - 1???)

Indeed, we actually correctly catch this result with `(q + 1)` as the midpoint (because we see that `q + 1` is a non-strict tupper bound for `p + size`).

We won't, however, catch an example like:
```cpp
void deref_plus_one(char* q) {
    char a = *(q + 1);
}

void test(unsigned size) {
    char *p = malloc(size);
    char *q = p + size - 1;
    deref_plus_one(q);
}
```
because the range-analysis library is intraprocedural, so we find `p + size` as the only expression that upper bounds `p + size`.

Reply:
3e6576bfaf131478b4a942f27e6e36acbc82a4e1 adds the above testcase.

Reply:
and 02076074ffa1fd8286c3db0fd5d9b8a9e718bf0a adds more comments to make the use of flow states more clear.

Reply:
1. Ah yes, I can see that the delta is connected between the source and the sink in this configuration.

2. Thanks for clearing this up.